packages/edit-widgets/src/utils/with-wp-customize.js

```diff
@@ -0,0 +1,36 @@
+/**
+ * WordPress dependencies
+ */
+import { serialize } from '@wordpress/blocks';
+
+/* global wp */
+
+function previewBlocksInWidgetArea( id, blocks ) {
+	const customizePreviewIframe = document.querySelector(
+		'#customize-preview > iframe'
+	);
+
+	if ( customizePreviewIframe && customizePreviewIframe.contentDocument ) {
+		const widgetArea = customizePreviewIframe.contentDocument.querySelector(
+			`[data-customize-partial-placement-context*='"sidebar_id":"${ id }"']`
+		);
+
+		if ( widgetArea ) {
+			widgetArea.innerHTML = serialize( blocks );
```

Comment:
I think this a great simple first step that allows us to try the block preview and is useful for testing with users.
Unfortunately, this method does not work with dynamic blocks. I guess we can use the ServerSideRender component to have a preview for these blocks.
Maybe we can serialize these blocks differently, e.g., a container div, and then we call React render on that div to render a ServerSideRender component?
I guess if we get this method working with dynamic blocks, we will be able to have widget previews as they are a dynamic block.

Reply:
That sounds like the way to go. I looked into the Customizer PHP API for this, but it is very opinionated on the format of the data that gets sent to the previewer. It looks like it expects widget form data.

Reply:
We could have an async serializer that uses `ServerSideRender` to render dynamic blocks. Should I go ahead with that?

@aduth I think you know a lot about the serializer right?

Reply:
I guess we can leave the preview of dynamic blocks for the second step in a follow-up. The objective of this phase is to have something users can test, and I think dynamic blocks are not a blocker. What do you think @youknowriad?

If we follow the logic of an async serializer it would be even more important that we don't serialize a block again if no changes happened. For each server-side rendered block we may have database queries and we should avoid making a server request each second. Maybe the problem is as simple as applying memoization on the serialization? (we would memoize based on block uid and the attributes object, so if the attributes did not change we would use the previous serialization result without any new request /rerender, not used memoization keys during serialization would be removed to avoid memory leaks or maybe a weakmap can also help us here).

Reply:
Yeah, memoization would do the trick.